docs: import export for project/env (#3369)
Unleash/unleash: website/docs/reference/deploy/environment-import-export.md

---
title: Environment Import & Export
---

:::info Availability

The environment import and export first appeared in Unleash 4.22.0.

:::

The environment export/import functionality in Unleash simplifies the process of moving feature flags between multiple instances. 
It also makes it easier to move feature flags within a single instance between environments.
When exporting you always select the toggles and environment to export from. When importing you always select a project and an environment to import to.

## Import & Export items

The following items will be exported and imported:
* feature toggles
* activation strategies
* variants

> * enabled/disabled statuses

Author: yes

> The following items will be exported and imported:
> * feature toggles
> * activation strategies
> * variants
> * enabled/disabled statuses
> * context fields
> * feature tags
> * feature tag types
> * segments

Reviewer: I think we could clarify already at this point that exporting features also exports global reference items etc.

we should also add doc links for as many items as we can on this list (I've not done that in the suggestion)

Here's is a rough suggestion for what I think would be good here. What do you think?

```suggestion
When you export features, the export will contain both feature-specific configuration and global configuration.

On the project-level these  items are exported:

* the feature itself
* feature tags

On the environment-level, these items are exported for the chosen environment:

* activation strategies including constraints and references to segments
* variants
* enabled/disabled

Additionally, these global configuration items are exported:
* custom context fields 
* feature tag types

Importantly, while references to segments are exported, the segments themselves are **not** exported. Consult the [import rejection](#import-rejection) section for more information.
```

Author: I like the extra clarity that it brings.

## Export

> Export is search-driven based on the results you see in the project toggle search or a global toggle search.

Reviewer: As a reader: what does search-driven mean? Is it something like this:

```suggestion
You can export features either from a project search or the global feature search. Use the search functionality to narrow the results to the list of features you want to export and use the export button to select environment. All features included in your search results will be included in the export.
```

But still, this feels very UI-centric and may be more at home in a how-to guide. How does the API work?

Author: This feature is UI centric as UI import is our main use case. I took the wording from our import/export article.

Author: At the API level you have a list of feature IDs to export and the env name

> ![Search-driven export screen with WYSIWYG mode](/img/export.png)

Reviewer: What is this image supposed to tell the reader? What information are they supposed to get from it? Just thinking the alt text could be better. And if it's a lot of words, consider using the `Figure` element and giving it a caption (check out `website/docs/topics/the-anatomy-of-unleash.mdx` for how to import and use the element).

## Import

Import is a 3 stage process designed to be efficient and error-resistant.

### Import stages

> * **upload** - you can upload exported file or copy-paste export data into the code editor

Reviewer: What's the file format? what about copy-pasted data? Must it be JSON or can it be ... XML? TOML? Where do they get this data? Is that copied from the exported file?

* **validation** - you will get feedback on any errors or warnings before you do the actual import. This makes sure your feature flags configurations
are compatible with your target environment.
* **import** - the actual import that creates a new configuration in the target environment or creates a [change request](../change-requests.md) when the environment has change requests enabled

![Three-staged import process](/img/import.png)

### Import rejection

> Import can be rejected when:

Reviewer: "Can" or "will"? 
```suggestion
An import will be rejected if:
```

> * the exported and imported context fields have conflicting context fields values. If the context fields don't exist in the target environment they will be created.

Reviewer: > the exported and imported context fields have conflicting context fields values.

What does this mean?

Author: Added clarification. Please let me know if it makes more sense now

* the target Unleash instance doesn't have a matching segment. The actual constraints in the segment can be different. The matching is done by the segment name.
* the exported strategies include custom strategies that don't exist in the target Unleash instance. The matching is done by the strategy name.

> * the exported feature already exists in another project in the target environment that is not a project that you import the toggle to.

Reviewer: ```suggestion
* you're trying to import a feature that already exists in a different project on the target instance.
```

Author: I like it

> * the target project and environment have a pending change request (applies to enterprise customers with change requests enabled for a given project and environment).

Reviewer: I think each list item here can be its own subsection so that we can explain it in full without overcrowding the list. There is a lot of nuance to some (most?) of the items and I don't think the list format serves that purpose very well.

### Import warnings

Import validation will warn you when:

> * the features you are trying to import are already archived in the target instance. Those features won't be imported again.

Reviewer: This is regardless of the what project they are archived from? If you're trying to import feature F into project P, but a feature called F has already been archived in project Q: what happens? Do you get an error or a warning?

Author: Added a tiny clarification "in any project". Isn't unique toggle per unleash instance documented somewhere else? Import/export is not changing anything in this regard.

> * the strategies you are trying to import are custom and may have different configuration parameters. Those strategies will be imported anyway.

Reviewer: > the strategies you are trying to import are custom and may have different configuration parameters.

What does it mean that they **may** have different parameters? Do you get this warning whenever you import custom strategies or only when you import custom strategies that have different parameters?

Author: Always when importing custom strategies. To remind you to check the strategy configuration if it happens to be part of our import.

Author: Added a link to configuration parameters

### Required permissions

> Import requires those permissions:

Reviewer: Judging from the list (and from what we discussed previously), it doesn't always need all of these, right? Sometimes, it's just a few of them?

> Import requires those permissions:
> * Update feature toggles (always)
> * Create feature toggles when import will be creating new features
> * Update tag types when import will be introducing new tag types
> * Create context fields when import will be creating new context fields
> * Create activation strategies when import will be creating new activation strategies and change requests are disabled
> * Delete activation strategies when import will be deleting existing activation strategies and change requests are disabled
> * Update variants when import will be updating variants and change requests are disabled

Reviewer: What happens to the last three when change requests are enabled? Does enabling change requests change any of the other ones?

Also is create/delete activation strategies related to creating/deleting custom strategies or to adding/removing strategies from a feature?

```suggestion
To import features, you will **always** require the **update feature toggles** permission.

Additionally, depending on how the actions your import job would trigger, you may also require any of the following permissions:

* **Create feature toggles**: When the import would create new features
* **Update tag types**: when the import would create new tag types
* **Create context fields**: when the import would create new context fields
* **Create activation strategies**: when the import would add (or is this "create custom activation strategies"? custom or standard?) activation strategies to a feature and change requests are disabled
* **Delete activation strategies**: when import would remove existing activation strategies from a feature and change requests are disabled
* **Update variants**: when the import would update variants and change requests are disabled

When change requests are enabled ....
```

Author: I like the suggestion

### Import and change requests

When the target project and environment have change requests enabled the import will not be applied instantly. Instead, it will 
create a change request, send it to review and await for the approval.

> 

Reviewer: Maybe we could add a paragraph here explaining how it affects permissions?

```suggestion
If change requests are enabled, any permissions for .... are not required.
```

Author: Agree, Let's be explicit and add this here

## Environment Import & Export vs Old Import & Export

The new Environment Import & Export differs from the old [Import & Export](import-export.md).
The most important difference is a strong focus on a single project and environment in the target environment.

> Also, the new Import & Export has a UI and extra safety mechanisms preventing corrupted data imports.

Reviewer: This is nice and a good clarification. So when would you pick the new one over the old one? And maybe more interestingly, when would you pick the old one over the new one?

Author: Hmm. This is on line 80 and 81. When you want to import into single project and environment. And if you want to have a UI and the validation stage.

The motivation for the old one should be in the docs of the old one I assume. From what I know it was trying to be too ambitious and export too much, but you'd need to talk to @ivarconr
